### PR TITLE
A number of bug fix and improvements

### PR DESCRIPTION
**Pull request by piiswrong**

### CODE REVIEW

**Comment by tqchen** [comment]
- It seems to me that this op can be separated into two ops, sum op and elementwise product op. 
- Please add comment on the operator itself, also try to make

**Comment by piiswrong** [comment]
This takes less memory

**Comment by piiswrong** [comment]
@tqchen How about a ADDITIONAL_OPERATOR_DIR flag for including custom operators in another folder?

**Comment by tqchen** [comment]
This is a great idea

**Comment by piiswrong** [comment]
@tqchen do you know how to modify the makefile for that?

**Comment by tqchen** [comment]
Something like https://github.com/dmlc/mxnet/blob/master/Makefile#L96 to add additional path to linker. Can also consider build the additional operators as independent so, and optionally load them in python.

**Comment by piiswrong** [comment]
@tqchen

**Comment by tqchen** [comment]
@piiswrong Let us consider not put softmax in Activation, but simply add another SoftmaxTransform operator as discussed in #472

**Comment by tqchen** [comment]
The major reason for not posing Softmax as an activation, is that it is kinda of special and do not fit into the template of normal operators. We need to add:

- add  mshadow support for general softmax gradient
- if you think we should place this on activation, let us open a issue to vote on this.

For now, can we isolate the change to the softmaxTransform out and merge other changes?

**Comment by piiswrong** [comment]
Conceptually I'm OK with either way. But an independent softmax involves a lot of code duplication

**Comment by mli** [comment]
`tools/caffe_converter/README.md` is also slightly updated in #969

**Comment by piiswrong** [comment]
@mli Ok, I have rebased

File: src/operator/softmax_activation.cc [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+// DO_BIND_DISPATCH comes from operator_common.h
+Operator *SoftmaxActivationProp::CreateOperator(Context ctx) const {
+  DO_BIND_DISPATCH(CreateOp, param_);
+}
+
+DMLC_REGISTER_PARAMETER(SoftmaxActivationParam);
+
+MXNET_REGISTER_OP_PROPERTY(SoftmaxActivation, SoftmaxActivationProp)
```
Comment (tqchen) [inline src/operator/softmax_activation.cc]: Consider use the name SoftmaxTransform?

Comment (piiswrong) [inline src/operator/softmax_activation.cc]: transform isn't used anywhere else in the code base. It's meaning is kind of unclear

Comment (tqchen) [inline src/operator/softmax_activation.cc]: OK， let us go with this one then.

File: src/operator/cudnn_activation-inl.h [hunk 2]
```diff
@@ -57,8 +57,19 @@ class CuDNNActivationOp : public Operator {
       data = in_data[activation::kData].get_with_shape<gpu, 4, real_t>(dshape, s);
       out = out_data[activation::kOut].get_with_shape<gpu, 4, real_t>(dshape, s);
     } else {
-      data = in_data[activation::kData].get<gpu, 4, real_t>(s);
-      out = out_data[activation::kOut].get<gpu, 4, real_t>(s);
+      Shape<4> dshape;
+      index_t size_left = in_data[activation::kData].Size();
```
Comment (tqchen) [inline src/operator/cudnn_activation-inl.h]: Is this still needed? as softmax is already isolated out?

**Comment by tqchen** [comment]
The code LGTM. Some final comments:
- I guess maybe cudnn_activation do not need to be touched as all the softmax related code are moved to SoftmaxActivation
- The default optimizer should remain as sgd, mainly because ccsgd seems is not able to be touched by learning rate scheduler

File: src/operator/cudnn_activation-inl.h (hunk 2 above)
Comment (piiswrong) [inline src/operator/cudnn_activation-inl.h]: these are bug fixes for cudnn_activation. It used to only accept 2 or 4 dimensional input.

**Comment by piiswrong** [comment]
ccsgd does use learning rate scheduler. Learning rate is pass into c++ side at every update

**Comment by tqchen** [comment]
OK, please do a squash and merge it in
